### PR TITLE
feat(integrations): dropbox

### PR DESCRIPTION
**Pull request by chloedia**
# Description

## Checklist before requesting a review

Please delete options that are not relevant.

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented hard-to-understand areas
- [ ] I have ideally added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged

### CODE REVIEW

File: backend/api/quivr_api/main.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 from quivr_api.routes.subscription_routes import subscription_router
 from sentry_sdk.integrations.fastapi import FastApiIntegration
 from sentry_sdk.integrations.starlette import StarletteIntegration
+from starlette.middleware.sessions import SessionMiddleware
```
Comment (StanGirard) [inline backend/api/quivr_api/main.py:31]: to remove?

File: backend/api/quivr_api/main.py [hunk 2]
```diff
@@ -70,6 +71,7 @@ def before_send(event, hint):
 
 app = FastAPI()
 add_cors_middleware(app)
+app.add_middleware(SessionMiddleware, secret_key=str(os.getenv("SESSION_SECRET_KEY")))
```
Comment (StanGirard) [inline backend/api/quivr_api/main.py:74]: same ?

File: backend/api/quivr_api/modules/sync/controller/dropbox_sync_routes.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        raise HTTPException(status_code=400, detail="Invalid state parameter")
+    else:
+        logger.info(
+            f"CURRENT USER: {current_user}, SYNC USER STATE USER: {sync_user_state.get('user_id')}"
+        )
+
+    if sync_user_state.get("user_id") != current_user:
+        raise HTTPException(status_code=400, detail="Invalid user")
+
+    redirect_uri = f"{BASE_REDIRECT_URI}"
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/controller/dropbox_sync_routes.py:119]: not needed ?

File: backend/api/quivr_api/modules/sync/utils/list_files.py [hunk 4]
```diff
@@ -292,3 +295,127 @@ def get_azure_files_by_id(credentials: dict, file_ids: List[str]):
...
+        recursive (bool, optional): If True, fetch files from all subfolders. Defaults to False.
+
+    Returns:
+        dict: A dictionary containing the list of files or an error message.
+    """
+    logger.info("Retrieving Dropbox files with folder_id: %s", folder_id)
+
+    # Verify credential has the access token
+    if "access_token" not in credentials:
+        raise HTTPException(status_code=401, detail="Invalid token data")
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/list_files.py:319]: HTTPException should only appear at the HTTP layer (routes) of the app. This function should raise some Exception (to define) and catched on the layer above

File: backend/api/quivr_api/modules/sync/utils/list_files.py [hunk 5]
```diff
@@ -292,3 +295,127 @@ def get_azure_files_by_id(credentials: dict, file_ids: List[str]):
...
+
+        logger.info("Dropbox files retrieved successfully: %d", len(files))
+        return files
+
+    except dropbox.exceptions.ApiError as e:
+        logger.error("Dropbox API error: %s", e)
+        raise HTTPException(status_code=500, detail="Dropbox API error")
+    except Exception as e:
+        logger.error("Unexpected error: %s", e)
+        raise HTTPException(status_code=500, detail="Unexpected error occurred")
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/list_files.py:369]: Same as previous comment on HTTP Exceptions

File: backend/pyproject.toml [hunk 6]
```diff
@@ -10,6 +10,8 @@ readme = "README.md"
 [tool.poetry.dependencies]
 python = "^3.11"
 
+dropbox = "^12.0.2"
+itsdangerous = "^2.2.0"
```
Comment (AmineDiro) [inline backend/pyproject.toml:14]: I don't see the use of this package 🤔 ? was this needed somewhere ?

File: backend/api/quivr_api/modules/sync/utils/dropboxutils.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+            dict: A dictionary containing the status of the download or an error message.
+        """
+        dbx = dropbox.Dropbox(token_data["access_token"])
+        dbx.check_and_refresh_access_token()
+        token_data["access_token"] = dbx._oauth2_access_token
+
+        downloaded_files = []
+        bulk_id = uuid.uuid4()
+        for file in files:
+            logger.info("🔥🔥🔥🔥: %s", file)
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/dropboxutils.py:73]: I hate to do remove fire emojies 😢  but we likely need to parse or search logs so a more structured logging is needed.

File: backend/api/quivr_api/modules/sync/utils/dropboxutils.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+                    logger.info("File is not compatible: %s", file_name)
+                    continue
+
+                to_upload_file = UploadFile(
+                    file=file_data,
+                    filename=file_name,
+                )
+
+                # Check if the file already exists in the database
+                existing_files = self.sync_files_repo.get_sync_files(sync_active_id)
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/dropboxutils.py:124]: I think that this should be refactored. Checking if a file exists should be a method that takes in a file and returns a boolean.

File: backend/api/quivr_api/modules/sync/utils/dropboxutils.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+                existing_file = next(
+                    (f for f in existing_files if f.path == file_name), None
+                )
+
+                supported = False
+                if (existing_file and existing_file.supported) or not existing_file:
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/dropboxutils.py:130]: I see duplicate logic on the if/else statements. Is there a way to have a more compact branching logic here 🤔

File: backend/api/quivr_api/modules/sync/utils/dropboxutils.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+                    supported = True
+                    await upload_file(to_upload_file, brain_id, current_user, bulk_id)
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/dropboxutils.py]: Probably needs refactoring. We already have downloaded the file in memory so we should just write it somewhere. The wrapper of `UploadFile` which is a FastApi struct is unnecessary IMHO

File: backend/api/quivr_api/modules/sync/utils/dropboxutils.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+                            syncs_active_id=sync_active_id,
+                            last_modified=modified_time,
+                            brain_id=brain_id,
+                            supported=supported,
+                        )
+                    )
+
+                downloaded_files.append(file_name)
+
+            except Exception as error:
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/utils/dropboxutils.py:165]: Could probably divide function into two.The Exception we catch is too generic in my sense and the block after the exception has a lot of logic that could be not related to the exception raised.  If the logic after the exception can be move before doing any logic, maybe it's better to have it upfront.

File: backend/api/quivr_api/modules/sync/controller/dropbox_sync_routes.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        user_id=str(current_user.id),
+        provider="DropBox",
+        credentials={},
+        state={"state": state},
+    )
+    sync_user_service.create_sync_user(sync_user_input)
+    return {"authorization_url": authorize_url}
+
+
+# {'state': 'dNzjXyh7CuAFyrBhKNG-UQ==|user_id=39418e3b-0258-4452-af60-7acfcc1263ff, name=chloe'}
```
Comment (StanGirard) [inline backend/api/quivr_api/modules/sync/controller/dropbox_sync_routes.py:77]: remove

File: backend/api/quivr_api/main.py (hunk 2 above)
Comment (chloedia) [reply backend/api/quivr_api/main.py:74]: Not working if not here at the moment

File: backend/pyproject.toml (hunk 6 above)
Comment (chloedia) [reply backend/pyproject.toml:14]: It is for the SessionMiddleware

**Pull request merged**